src/cloud-api/src/client.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! # Materialize cloud API client
+//!
+//! This module provides an API client with typed methods for
+//! interacting with the Materialize cloud API. This client i,
+//! token management, and basic requests against the API.
+//!
+//! The [`Client`] requires an [`mz_frontegg_client::client::Client`] as a parameter. The
+//! Frontegg client is used to request and manage the access token.
+
+use reqwest::{Method, RequestBuilder, Url};
+use serde::{de::DeserializeOwned, Deserialize};
+
+use crate::error::{ApiError, Error};
+
+/// Represents the structure for the client.
+#[allow(dead_code)]
+pub struct Client {
+    pub(crate) inner: reqwest::Client,
+    pub(crate) auth_client: mz_frontegg_client::client::Client,
+    pub(crate) endpoint: Url,
+}
+
+pub mod cloud_provider;
+pub mod environment;
+pub mod region;
+
+/// Cloud endpoints architecture:
+///
+/// (CloudProvider)                         (Region)                                 (Environment)
+///   ---------              --------------------------------------            ------------------------
+///  |          |           |          Region Controller           |          | Environment Controller |
+///  |  Cloud   |  api_url  |    ----------        -------------   |  ec_url  |                        |
+///  |  Sync    | --------> |   | Provider | ---- |    Region   |  | -------> |       Environment      |
+///  |          |           |   | (aws..)  |      |  (east-1..) |  |          |  (pgwire_address...)   |
+///  |          |           |    ----------        -------------   |          |                        |
+///   ----------             --------------------------------------            -----------------------
+///
+impl Client {
+    /// Builds a request towards the `Client`'s endpoint
+    async fn build_request<P>(
```

Comment:
Minor nit: I'd add two helper methods. One called `build_global_request` that calls this with `sync` as the subdomain, and one called `build_region_request` that takes a `Region` as an argument and then calls this with `region.ec_subdomain` as the subdomain. Will cut down on a little boilerplate at the callsites!

Reply:
And looking at it a little more, I think you want this method to be `domain: &str`, so that you don't have to muck around with subdomains in the codepaths that don't want to deal in subdomains.